lib/tpm_policy.c

```diff
@@ -0,0 +1,254 @@
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//**********************************************************************;
+#include <errno.h>
+#include <limits.h>
+
+#include "../lib/tpm2_util.h"
+#include "files.h"
+#include "log.h"
+#include "tpm_session.h"
+#include "tpm_hash.h"
+#include "tpm_policy.h"
+
+static unsigned get_size_from_alg(TPMI_ALG_HASH hashAlg) {
+    switch (hashAlg) {
+        case TPM_ALG_SHA1:
+            return SHA1_DIGEST_SIZE;
+            break;
+        case TPM_ALG_SHA256:
+            return SHA256_DIGEST_SIZE;
+            break;
+        case TPM_ALG_SHA384:
+            return SHA384_DIGEST_SIZE;
+            break;
+        case TPM_ALG_SHA512:
+            return SHA512_DIGEST_SIZE;
+            break;
+        case TPM_ALG_SM3_256:
+            return SM3_256_DIGEST_SIZE;
+            break;
+        default:
+            return SHA1_DIGEST_SIZE;
+            break;
+    }
+}
+
+static bool evaluate_populate_pcr_digests(CREATE_POLICY_CTX *pctx, TPML_DIGEST *pcr_values) {
+    
+    uint8_t group_val=0;
+    uint8_t total_indices_for_this_alg=0;
+    uint8_t expected_pcr_input_file_size=0;
+    unsigned i, j, k, dgst_cnt=0; 
+    for (i=0; i<pctx->pcr_policy_options.pcr_selections.count; i++) { 
+        for (j=0; j<pctx->pcr_policy_options.pcr_selections.pcrSelections[i].sizeofSelect; j++) {
+            group_val = pctx->pcr_policy_options.pcr_selections.pcrSelections[i].pcrSelect[j];
+            total_indices_for_this_alg+=(((group_val>>3)&1)+((group_val>>2)&1)+((group_val>>1)&1)+(group_val&1));
+        }
+        expected_pcr_input_file_size += (total_indices_for_this_alg * 
+            get_size_from_alg(pctx->pcr_policy_options.pcr_selections.pcrSelections[i].hash));
+        /*
+         * indexing with k is used as valid map while populating from input file if 
+         * file_size_test passes
+         */
+        pcr_values->count += total_indices_for_this_alg;
+        for (k=0; k<total_indices_for_this_alg; k++) {
+            pcr_values->digests[dgst_cnt+k].t.size = 
+                get_size_from_alg(pctx->pcr_policy_options.pcr_selections.pcrSelections[i].hash);
+        }
+        dgst_cnt++;
+
+        total_indices_for_this_alg=0;
+    }
+
+    int file_size_test = (pcr_values->count == 0) ? 1:0;
```

Comment:
No need for this, just use pcr_values->count == 0, since that provides you a 1 via language guarantees.